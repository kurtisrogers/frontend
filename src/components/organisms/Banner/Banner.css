.banner {
  position: relative;
  align-items: center;
  height: 100vh;
}
