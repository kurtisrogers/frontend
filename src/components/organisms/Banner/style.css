.banner {
  display: grid;
  width: 100%;
  grid-auto-flow: column;
  grid-auto-columns: 1fr;
  grid-auto-rows: min-content;
  /* overflow-x: clip; */
  gap: 1rem;
}

.banner--image {
  display: grid;
  width: 100%;
  min-height: calc(100vh - 53.5px);
  grid-auto-flow: column;
  grid-auto-columns: 1fr;
  grid-auto-rows: min-content;
  /* overflow-x: clip; */
  gap: 1rem;
  align-content: center;
}

.banner .banner__content {
  grid-column: span 14;
  align-self: center;
  z-index: 1;
}

.banner--image .banner__content {
  grid-column: span 6;
  align-self: center;
  z-index: 1;
}

.banner__content ul {
  padding-left: 1em;
}

.banner__content h1 {
  margin-bottom: 0px;
}

.banner__image {
  grid-column: span 8;
}

.banner--image img {
  display: block;
  object-fit: cover;
  width: 130%;
  height: calc(100vh - 107px);
  margin-left: -30%;
}