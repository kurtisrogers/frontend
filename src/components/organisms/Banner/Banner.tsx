import { Show } from 'solid-js';
import type { Image } from '@/types/branding';
import "./Banner.css";

interface Props {
  title?: string;
  backgroundImage?: Image;
};

export default function Banner( props: Readonly<Props> ) {
  const { title, backgroundImage } = props;

  if (!title) return;

  return (
    <div class="banner content-grid">
      <h1>{title}</h1>
      {/* weird image/background colour section */}
      <Show when={backgroundImage}>
        <div class="banner__image">
          <img src={backgroundImage?.src} alt={backgroundImage?.alt} />
        </div>
      </Show>
    </div>
  )
};
