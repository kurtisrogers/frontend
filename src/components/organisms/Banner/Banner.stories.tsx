import Banner from "./index";
import "./style.css";

export default {
  title: "Organisms/Banner",
  component: Banner,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/solid/writing-docs/docs-page
  tags: ["autodocs"],
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/solid/configure/story-layout
    layout: "fullscreen"
  }
};

const propData = {
  title: "Example",
  children: (
    <>
      <p>My name is Kurtis Rogers, a proper Software Engineer from Bristol, UK</p>
      <ul>
        <li>Specialising in Frontend technology since 2018</li>
        <li>Expert in leading frameworks including serverside Node development (Express/Koa)</li>
        <li>
          Experienced in building and developing accessible themes in both Drupal and WordPress
        </li>
        <li>Effective and approachable mentor</li>
        <li>
          Team player, able to build robust yet relaxed relationships with colleagues and
          stakeholders alike
        </li>
      </ul>
      <p>Want to know more?</p>
      <a class="button" href="/about">
        Read more about me
      </a>
    </>
  ),
  backgroundImage: {
    data: {
      attributes: {
        alternativeText:
          "A photo of The Milky Way, lots of stars and some star-making gas clouds far far away...",
        url: "/images/pexels-jack-davis-86003658-9069564-thumbnail.webp",
        formats: {
          thumbnail: {
            src: "/images/pexels-jack-davis-86003658-9069564-thumbnail.webp"
          },
          small: {
            src: "/images/pexels-jack-davis-86003658-9069564-sm.webp",
            minWidth: 320
          },
          medium: {
            src: "/images/pexels-jack-davis-86003658-9069564-md.webp",
            minWidth: 768
          },
          large: {
            src: "/images/pexels-jack-davis-86003658-9069564-lg.webp",
            minWidth: 960
          },
          xlarge: {
            src: "/images/pexels-jack-davis-86003658-9069564-xl.webp",
            minWidth: 1420
          }
        }
      }
    }
  }
};

const Template = () => {
  return (
    <Banner title={propData.title} backgroundImage={propData.backgroundImage}>
      {propData.children}
    </Banner>
  );
};

export const Default = Template.bind({});
// Default.args = {
//   // Add any props you want to pass to the Header component
// };
