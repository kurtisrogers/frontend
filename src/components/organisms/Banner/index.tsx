import { Show, For } from "solid-js";
import type { ImageResponse } from "@/types/branding";
import "./style.css";

import responsiveImages from "@/helpers/responsiveImages";
import { Dynamic } from "solid-js/web";
import { LayoutSpacingDataType } from "@/helpers/layoutSpacingHandler";
import type { Heading } from "@/types/branding";
import renderList from "@/helpers/renderList";
import Button from "@/components/atoms/Button";

export interface Props {
  title?: Heading | string;
  image: {
    data: ImageResponse;
  };
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  children: any[];
  firstChild?: boolean;
  spacing?: LayoutSpacingDataType;
}

export default function Banner(props: Readonly<Props>) {
  const { title, image, children, firstChild = "dark" } = props;
  const { data: { attributes: { alternativeText = "", credit } = {} } = {} } = image ?? {};

  return (
    <section class={`banner ${firstChild ? "parent-section" : "child-section"}`}>
      <div class={`banner__content`}>
        <hgroup class={`banner__content--copy fade--in_up`}>
          <Show when={title}>
            <Dynamic
              component={`h${typeof title === "string" ? "1" : title?.headingLevel}`}
              class={typeof title === "string" ? "text-size-h1" : title?.headingClass}
            >
              {typeof title === "string" ? title : title?.text}
            </Dynamic>
          </Show>
          <For each={children}>
            {child => {
              const Content = renderList[child?.type as keyof typeof renderList];
              return <Content {...child} />;
            }}
          </For>
        </hgroup>
        {/* TODO: weird image/background colour section - not fully made my mind up about this */}
        {/* TODO: create an image component - generate images here: https://isrcset.com/generate */}
        <div class="banner__content--image fade--in">
          <Show when={image}>
            {responsiveImages(image)}
            <Show when={alternativeText}>
              <div class="credit">
                {/* TODO: will require work on the backend extending the existing image controller */}
                <Show when={credit}>
                  <Button href={credit?.link}>
                    <span class="sr-only">Photo by </span>
                    {credit?.author}
                  </Button>
                </Show>
                <span>{alternativeText}</span>
              </div>
            </Show>
          </Show>
        </div>
      </div>
    </section>
  );
}
