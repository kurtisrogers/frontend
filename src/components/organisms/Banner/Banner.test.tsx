import { describe, expect, vi, beforeEach, afterEach, it } from "vitest";
import { render } from "@solidjs/testing-library";
import { screen } from "@testing-library/dom";
import Banner from ".";

describe("Banner", () => {
  beforeEach(() => {
    // Mock the fetch function
    global.fetch = vi.fn();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  const propData = {
    type: "banner",
    title: {
      text: "About",
      headingLevel: "1" as "1" | "2" | "3" | "4" | "5" | "6"
    },
    children: [
      {
        type: "paragraph",
        children: [
          {
            type: "text",
            text: "Technology is where I live. I also love anything to do with space."
          },
          {
            type: "text",
            text: "Will our species ever reach Alpha Centauri? Propulsion technology has come a long way since we last went to the Moon. A simple laser-powered lightsail? Theoretically this could reach 20% speed of light (0.2c). So Alpha Centauri in a cool 21.41 years. Not sure about the return journey though..."
          },
          {
            type: "text",
            text: "Time dialation? 0.44 years. And roughly half an hour away via radio contact..."
          }
        ]
      }
    ],
    image: {
      data: {
        attributes: {
          alternativeText: "Milky Way illustration.",
          credit: {
            author: "Philippe Donn",
            link: {
              url: "https://www.pexels.com/photo/milky-way-illustration-1169754/",
              target: "_blank" as "_blank" | "_self"
            }
          },
          url: "/images/optimised/milkyway-galaxy-md.webp",
          formats: {
            medium: {
              src: "/images/optimised/milkyway-galaxy-md.webp",
              minWidth: 768
            },
            large: {
              src: "/images/optimised/milkyway-galaxy-lg.webp",
              minWidth: 960
            },
            xlarge: {
              src: "/images/optimised/milkyway-galaxy-xl.webp",
              minWidth: 1420
            }
          }
        }
      }
    }
  };

  it("renders", () => {
    const { container } = render(() => <Banner {...propData} />);
    expect(container).toMatchSnapshot();
  });

  it("has a title", () => {
    const { getByText } = render(() => <Banner {...propData} />);

    expect(getByText(/About/)).toBeInTheDocument();
  });

  it("can have a image", () => {
    render(() => <Banner {...propData} />);

    // Find the image inside the picture element
    const image = screen.getByAltText(/Milky Way illustration./i);

    // Check if the image exists and has the correct alt text
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("src", propData.image.data.attributes.url);
  });
});
