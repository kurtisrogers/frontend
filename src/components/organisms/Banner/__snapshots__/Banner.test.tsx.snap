// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Banner > renders 1`] = `
<div>
  <section
    class="banner parent-section"
  >
    <div
      class="banner__content"
    >
      <hgroup
        class="banner__content--copy fade--in_up"
      >
        <h1>
          About
        </h1>
        <p>
          Technology is where I live. I also love anything to do with space.
        </p>
        <p>
          Will our species ever reach Alpha Centauri? Propulsion technology has come a long way since we last went to the Moon. A simple laser-powered lightsail? Theoretically this could reach 20% speed of light (0.2c). So Alpha Centauri in a cool 21.41 years. Not sure about the return journey though...
        </p>
        <p>
          Time dialation? 0.44 years. And roughly half an hour away via radio contact...
        </p>
      </hgroup>
      <div
        class="banner__content--image fade--in"
      >
        <picture>
          <source
            media="(min-width: 1420px)"
            srcset="/images/optimised/milkyway-galaxy-xl.webp"
          />
           
          <source
            media="(min-width: 960px)"
            srcset="/images/optimised/milkyway-galaxy-lg.webp"
          />
           
          <img
            alt="Milky Way illustration."
            class="w-full h-full object-cover"
            decoding="async"
            loading="lazy"
            src="/images/optimised/milkyway-galaxy-md.webp"
          />
        </picture>
        <div
          class="credit"
        >
          <a
            class="bg-red text-white   "
            href="https://www.pexels.com/photo/milky-way-illustration-1169754/"
            target="_blank"
          >
            <span
              class="sr-only"
            >
              Photo by 
            </span>
            Philippe Donn
            <span
              class="sr-only"
            >
               (opens a new tab)
            </span>
          </a>
          <span>
            Milky Way illustration.
          </span>
        </div>
      </div>
    </div>
  </section>
</div>
`;
